apps/arweave/src/ar_mining_worker.erl

```diff
@@ -245,362 +284,412 @@ terminate(_Reason, _State) ->
-				{ok, PoA} ->
-					PoA;
+				{ok, PoA} -> PoA;
 				{error, _Error} ->
 					%% Fallback. This will probably fail later, but prepare_poa/3 should
 					%% have already printed several errors so we'll continue just in case.
+					%% df: Is this the right fallback?..
 					#poa{ chunk = Chunk2 }
 			end,
 			ar_coordination:computed_h2_for_peer(Candidate#mining_candidate{ poa2 = PoA2 })
 	end,
-	{noreply, State2};
+	%% Remove the cached value from the cache.
+	case ar_mining_cache:with_cached_value(
+		?VDF_STEP_CACHE_KEY(Candidate#mining_candidate.cache_ref, Candidate#mining_candidate.nonce),
+		Candidate#mining_candidate.session_key,
+		State1#state.chunk_cache,
+		fun(_) -> {ok, drop} end
+	) of
+		{ok, ChunkCache2} -> State1#state{ chunk_cache = ChunkCache2 };
+		{error, Reason} ->
+			?LOG_ERROR([{event, mining_worker_failed_to_process_computed_h2},
+				{worker, State1#state.name}, {partition, State1#state.partition_number},
+				{nonce, Candidate#mining_candidate.nonce},
+				{session_key, ar_nonce_limiter:encode_session_key(Candidate#mining_candidate.session_key)},
+				{reason, Reason}]),
+			State1
+	end;
 
+%% @doc Handle the `compute_h2_for_peer` task.
+%% Indicates that we got a request to compute h2 for a peer.
 handle_task({compute_h2_for_peer, Candidate, _ExtraArgs}, State) ->
 	#mining_candidate{
-		session_key = SessionKey,
 		h0 = H0,
 		partition_number = Partition1,
 		partition_upper_bound = PartitionUpperBound,
 		cm_h1_list = H1List,
-		cm_lead_peer = Peer,
-		packing_difficulty = PackingDifficulty
+		cm_lead_peer = Peer
 	} = Candidate,
-
-	{_RecallRange1Start, RecallRange2Start} = ar_block:get_recall_range(H0,
-					Partition1, PartitionUpperBound),
+	{_, RecallRange2Start} = ar_block:get_recall_range(H0, Partition1, PartitionUpperBound),
 	Candidate2 = generate_cache_ref(Candidate),
-	%% Clear the list so we aren't copying it around all over the place
+	%% Clear the list so we aren't copying it around all over the place.
 	Candidate3 = Candidate2#mining_candidate{ cm_h1_list = [] },
-
 	Range2Exists = ar_mining_io:read_recall_range(chunk2, self(), Candidate3, RecallRange2Start),
 	case Range2Exists of
 		true ->
 			ar_mining_stats:h1_received_from_peer(Peer, length(H1List)),
-			%% Note: when processing CM requests we always reserve the cache space and proceed
-			%% *even if* this puts us over the chunk cache limit. This may have to be
-			%% revisited later if we find that this causes unacceptable memory bloat.
-			State2 = update_sub_chunk_cache_size(
-				ar_block:get_nonces_per_recall_range(PackingDifficulty), SessionKey, State),
-			%% First flag all nonces in the range as do_not_cache, then cache the specific
-			%% nonces included in the H1 list. This will make sure we don't cache the chunk2s
-			%% that are read for the missing nonces.
-			State3 = do_not_cache(Candidate3, State2),
-			{noreply, cache_h1_list(Candidate3, H1List, State3)};
+			State1 = mark_chunk2_missing(Candidate3, State),
```

Comment:
why do we mark chunk2 missing here? I'm still working to sort out the new cache state (which is much better than the confusing cycle_chunk_cache stuff) - I thought we only marked chunk2 missing if Range2 does not exist

Reply:
Yea, the logic here was that previously we would cache both chunks in the same place, assuming that chunk2 would replace chunk1. So previously we marked the whole range as `do_not_cache` and then, on the next line, we stored the H1s with `cache_h1_list`, replacing these `do_not_cache` flags with actual H1s; all the range was marked as `do_not_cache`, except for the chunks we have H1 for.
During rework that the logic changed, and now we mark the whole range as `chunk2_missing` (separate record flag), and then traverse the list of H1s, storing those we just received from peer. The only thing missing here is that we need to also reset the `chunk2_missing` flag to `false` for those entries we have H1 for. I'll fix it. 
Good call.